resources/todomvc/architecture-examples/lit/src/lib/todos.ts

```diff
@@ -0,0 +1,141 @@
+export interface Todo {
+    id: number;
+    text: string;
+    completed: boolean;
+}
+
+export type TodoEdit = Partial<Todo> & { id: number };
+
+const todoFilters = ["all", "active", "completed"] as const;
+export type TodoFilter = (typeof todoFilters)[number];
+function isTodoFilter(value: string | undefined): value is TodoFilter {
+    return todoFilters.includes(value as TodoFilter);
+}
+
+/**
+ * A mutable, observable container for a todo list.
+ *
+ * @fires a `change` event when the todo list changes.
+ */
+export class Todos extends EventTarget {
+    #nextId = 1;
+    #todos: Array<Todo> = [];
+    #filter: TodoFilter = this.#filterFromUrl();
+
+    get all(): ReadonlyArray<Todo> {
+        return this.#todos;
+    }
+
+    get active(): ReadonlyArray<Todo> {
+        return this.#todos.filter((todo) => !todo.completed);
+    }
+
+    get completed(): ReadonlyArray<Todo> {
+        return this.#todos.filter((todo) => todo.completed);
+    }
+
+    get allCompleted(): boolean {
+        return this.#todos.every((todo) => todo.completed);
+    }
+
+    connect() {
+        window.addEventListener("hashchange", this.#onHashChange);
+    }
+
+    disconnect() {
+        window.removeEventListener("hashchange", this.#onHashChange);
+    }
+
+    filtered() {
+        switch (this.#filter) {
+            case "active":
+                return this.active;
+            case "completed":
+                return this.completed;
+            case "all":
+            case undefined:
+            // fall through
+        }
+        return this.all;
+    }
+
+    #notifyChange() {
+        window.localStorage.setItem("todos", JSON.stringify({ todos: this.#todos, nextId: this.#nextId }));
```

Comment:
One more localStorage setter that can be removed here

Reply:
+1, done